### DISCUSSION TITLE
block subdomains?

### DISCUSSION THREAD

**Discussion by 4nd3r** [qutebrowser/qutebrowser#6340] [Show and tell] [answered]
Currently only explicit host in blocked hosts list is blocked.

With `hosts` files, Brave's Adblock and uBlock Origin block host and its subdomains.

1. Should [_is_blocked in hostblock.py](https://github.com/qutebrowser/qutebrowser/blob/fc93f1b06f8eb0c5e7774b6274dc576376cb655b/qutebrowser/components/hostblock.py#L114) behave same way?
2. Should Brave's Adblock, if available, also read `content.blocking.hosts.lists` and disable `hostblock.py`?

I think answer to first question should be yes and (quick'n'dirty) patch for that is rather simple:

```diff
diff --git a/qutebrowser/components/hostblock.py b/qutebrowser/components/hostblock.py
index 8a0174584..cbc370c9f 100644
--- a/qutebrowser/components/hostblock.py
+++ b/qutebrowser/components/hostblock.py
@@ -124,10 +124,16 @@ class HostBlocker:
         if not config.get("content.blocking.enabled", url=first_party_url):
             return False
 
+        if blockutils.is_whitelisted_url(request_url):
+            return False
+
         host = request_url.host()
-        return (
-            host in self._blocked_hosts or host in self._config_blocked_hosts
-        ) and not blockutils.is_whitelisted_url(request_url)
+
+        for blocked_host in self._blocked_hosts.union(self._config_blocked_hosts):
+            if host == blocked_host or host.endswith(f'.{blocked_host}'):
+                return True
+
+        return False
 
     def filter_request(self, info: interceptor.Request) -> None:
         """Block the given request if necessary."""
```

I don't know answer to second question, but related to that is also there is question about performance.

Which is faster - iterating over set/list/dict/whatever or throwing stuff at adblock?

**Comment by The-Compiler** [comment, marked as answer]
> Should _is_blocked in hostblock.py behave same way?

Not sure. It looks like (as you say), both uBO and the Brave adblocker interpret a rule like `example.com` as `||example.com^`, and indeed [that matches](https://adblockplus.org/filter-cheatsheet) something like `sub.example.com` as well. That's not, however, how hosts files work anywhere else, so it seems rather counter-intuitive to me.

@ArniDagur What do you think?

Also FWIW that patch is a no-go, as it iterates through all blocked hosts for every request. A possible implementation should probably use the `widened_hostnames` function which is used for the config code already, so that it can still benefit from simple set checking.

> Should Brave's Adblock, if available, also read content.blocking.hosts.lists and disable hostblock.py?

We briefly discussed that in #5317 and decided against it, to keep things simple. We'll need to keep the current code around anyways (as long as `adblock` isn't mandatory), and the non-adblock blocking method changing behavior depending on whether `adblock` is available or not seems weird.

> Which is faster - iterating over set/list/dict/whatever or throwing stuff at adblock?

Probably the latter, but in the current `hostblock.py` code there's no iteration over the blocked hosts.

**Reply by 4nd3r** [reply]
thank you for answering, really appreciate it.

> That's not, however, how hosts files work anywhere else, so it seems rather counter-intuitive to me.

That's true and then again often times blocking `*.example.com` is much more effective than listing all possible subdomains (and missing some). This is probably why `||example.com^` is interpreted this way.

Could this be (config) switchable feature for `hostblock.py`? There is undeniable benefit to it and decision stays with user.

And when enabled, we could even optimize blocked hosts list (after download) in a way that we exclude hosts which are already blocked by some other line. e.g remove `foo.bar.example.com`, because it's already blocked by `bar.example.com`. Tho, I think that this optimization wouldn't help that much in performance and just creates additional maintenance burden. Just thinking out loud 😄

**Reply by ArniDagur** [reply]
>> Should _is_blocked in hostblock.py behave same way?

> Not sure. It looks like (as you say), both uBO and the Brave adblocker interpret a rule like example.com as ||example.com^, and indeed that matches something like sub.example.com as well. That's not, however, how hosts files work anywhere else, so it seems rather counter-intuitive to me.

> @ArniDagur What do you think?

I don't have a strong opinion on the subject. I think it's fine that uBlock Origin / Brave block subdomains, but qutebrowser doesn't.

**Reply by The-Compiler** [reply]
I opened an issue for it: #6365. If someone comes up with an implementation which still relies on set membership (rather than iterating through all blocked domains for every request), I guess I'm fine with the change, but I'm unlikely to work on it myself.

**Comment by WheroTaniwha** [comment]
No idea, except that there are very strange things happening to my UblockOrigin. For example, I was sure I had made some changes in the settings but it appeared they had not been saved?? Anyhow, I called back the file from the last time I saved them and this came back .  . .  

{
  "timeStamp": 1617648001169,
  "version": "1.34.0",
  "userSettings": {
    "advancedUserEnabled": true,
    "cloudStorageEnabled": true,
    "contextMenuEnabled": false,
    "importedLists": [],
    "popupPanelSections": 31,
    "webrtcIPAddressHidden": true
  },
  "selectedFilterLists": [
    "user-filters",
    "ublock-filters",
    "ublock-badware",
    "ublock-privacy",
    "ublock-abuse",
    "ublock-unbreak",
    "adguard-mobile",
    "easylist",
    "adguard-spyware",
    "easyprivacy",
    "urlhaus-1",
    "adguard-annoyance",
    "adguard-social",
    "fanboy-thirdparty_social",
    "fanboy-cookiemonster",
    "fanboy-annoyance",
    "fanboy-social",
    "ublock-annoyances",
    "plowe-0"
  ],
  "hiddenSettings": {},
  "whitelist": [
    "about-scheme",
    "chrome-extension-scheme",
    "chrome-scheme",
    "edge-scheme",
    "moz-extension-scheme",
    "opera-scheme",
    "vivaldi-scheme",
    "wyciwyg-scheme"
  ],
  "dynamicFilteringString": "behind-the-scene * * noop\nbehind-the-scene * inline-script noop\nbehind-the-scene * 1p-script noop\nbehind-the-scene * 3p-script noop\nbehind-the-scene * 3p-frame noop\nbehind-the-scene * image noop\nbehind-the-scene * 3p noop",
  "urlFilteringString": "",
  "hostnameSwitchesString": "no-large-media: behind-the-scene false\nno-remote-fonts: * true\nno-large-media: * true\nno-csp-reports: * true",
  "userFilters": ""
}

. .  . Google seems to be working habd-in-glove with Microsoft Azure to knobble any and every possibility of users enabling even a smidgeon of privacy. Schemas schweeenah!
